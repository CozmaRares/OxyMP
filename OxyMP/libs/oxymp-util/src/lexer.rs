use regex::Regex;

pub enum TokenMatcher {
    Exact(String),
    Regex(Regex),
}

impl TokenMatcher {
    pub fn regex(re: &str) -> TokenMatcher {
        let re = format!("^{re}");
        let re = Regex::new(&re).unwrap();
        TokenMatcher::Regex(re)
    }
}

pub struct LexerState<'a> {
    input: &'a str,
    cursor: usize,
}

impl<'a> LexerState<'a> {
    fn advance(&mut self, n: usize) {
        self.cursor += n;
    }

    fn remaining(&self) -> &'a str {
        &self.input[self.cursor..]
    }

    fn has_remaining(&self) -> bool {
        self.cursor < self.input.len()
    }

    pub fn current_n(&self, n: usize) -> &'a str {
        &self.input[self.cursor..self.cursor + n]
    }

    pub fn current_offset(&self) -> usize {
        self.cursor
    }
}

type TokenCreatorFn<Token> = dyn Fn(&LexerState, usize) -> Token;
type TokenTransformerFn<Token> = dyn Fn(&LexerState, usize) -> Result<Token, LexError>;

pub enum TokenHandler<Token> {
    Pattern(Box<TokenCreatorFn<Token>>),
    Regex(Box<TokenTransformerFn<Token>>),
    Ignore,
}

pub struct LexRule<Token> {
    matcher: TokenMatcher,
    handler: TokenHandler<Token>,
}

impl<Token> LexRule<Token> {
    pub fn new(matcher: TokenMatcher, handler: TokenHandler<Token>) -> Self {
        LexRule { matcher, handler }
    }

    fn matches(&self, state: &LexerState) -> Option<usize> {
        match &self.matcher {
            TokenMatcher::Exact(exact_match) => state
                .remaining()
                .starts_with(exact_match)
                .then_some(exact_match.len()),
            TokenMatcher::Regex(re) => re
                .captures(state.remaining())
                .and_then(|captures| captures.get(0))
                .map(|matched| matched.end() - matched.start()),
        }
    }

    fn consume(&self, state: &LexerState, matched_size: usize) -> Result<Option<Token>, LexError> {
        match &self.handler {
            TokenHandler::Ignore => Ok(None),
            TokenHandler::Pattern(f) => Ok(Some(f(state, matched_size))),
            TokenHandler::Regex(f) => f(state, matched_size).map(|t| Some(t)),
        }
    }
}

#[derive(Debug)]
pub enum LexError {
    UnknownPattern(String),
    UnparsableToken(String),
}

impl LexError {
    fn unknown(input: &str) -> LexError {
        LexError::UnknownPattern(input.into())
    }

    pub fn unparsable(input: &str) -> LexError {
        LexError::UnparsableToken(input.into())
    }
}

impl std::fmt::Display for LexError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            LexError::UnknownPattern(input) => write!(f, "Unknown pattern: {}", input),
            LexError::UnparsableToken(input) => write!(f, "Unparsable token: {}", input),
        }
    }
}

impl std::error::Error for LexError {}

pub type LexResult<T> = Result<T, LexError>;

pub struct Lexer<Token> {
    pub rules: Vec<LexRule<Token>>,
}

impl<Token> Lexer<Token> {
    pub fn tokenize<'a>(&'a self, input: &'a str) -> Result<Vec<Token>, LexError> {
        let mut tokens = Vec::new();
        let mut state = LexerState { input, cursor: 0 };

        while state.has_remaining() {
            let mut was_consumed = false;

            for rule in &self.rules {
                let matched_size = match rule.matches(&state) {
                    None => continue,
                    Some(size) => size,
                };

                if let Ok(result) = rule.consume(&state, matched_size) {
                    if let Some(token) = result {
                        tokens.push(token);
                    }
                    was_consumed = true;
                    state.advance(matched_size);
                    break;
                }
            }
            if !was_consumed {
                return Err(LexError::unknown(input));
            }
        }
        Ok(tokens)
    }
}

#[cfg(debug_assertions)]
#[derive(Debug, Clone)]
pub struct TokenDebugInfo {
    pub inpup_offset: usize,
    pub matched_rule: String,
    pub matched: String,
    pub matched_size: usize,
}
