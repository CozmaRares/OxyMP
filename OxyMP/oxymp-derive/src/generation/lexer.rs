#![allow(non_snake_case)]

use quote::quote;

use crate::{
    data::MacroData,
    idents::{base_ident, tokens},
    symbols::{get_def, Symbol},
    tokens::{ExactToken, IgnorePattern, RegexToken, TokenInfo},
};

pub fn generate_constructor(data: &MacroData) -> proc_macro2::TokenStream {
    let visibility = &data.visibility;
    let token_info = &data.tokens;

    let _Rc = get_def(Symbol::Rc, data.simple_types);
    let _Box = get_def(Symbol::Box, data.simple_types);
    let _vec = get_def(Symbol::VecMacro, data.simple_types);
    let _LexRule = get_def(Symbol::UtilLexRule, data.simple_types);
    let _Lexer = get_def(Symbol::UtilLexer, data.simple_types);
    let _TokenHandler = get_def(Symbol::UtilTokenHandler, data.simple_types);
    let _TokenMatcher = get_def(Symbol::UtilTokenMatcher, data.simple_types);

    let rules = token_info.iter().map(|tok| match tok {
        TokenInfo::Exact(ExactToken { name, pattern }) => {
            let enum_ident = tokens::enum_ident(name);
            let struct_ident = tokens::struct_ident(name);

            quote! {
                #_LexRule {
                    matcher: #_TokenMatcher::Exact(#pattern.to_string()),
                    handler: #_TokenHandler::Pattern(Token::#enum_ident(#struct_ident))
                }
            }
        }
        TokenInfo::Regex(RegexToken {
            name,
            regex,
            transformer_fn,
            ..
        }) => {
            let enum_ident = tokens::enum_ident(name);
            let struct_ident = tokens::struct_ident(name);

            quote! {
                #_LexRule {
                    matcher: #_TokenMatcher::regex(#regex),
                    handler: #_TokenHandler::Regex(
                        #_Box::new(
                            |matched| Token::#enum_ident(#struct_ident(#_Rc::new(#transformer_fn(matched))))
                        )
                    )
                }
            }
        }
        TokenInfo::Ignore(IgnorePattern { regex }) => quote! {
            #_LexRule {
                matcher: #_TokenMatcher::regex(#regex),
                handler: #_TokenHandler::Ignore
            }
        },
    });

    quote! {
        #visibility fn create_lexer() -> #_Lexer<Token> {
            #_Lexer { rules: #_vec![#(#rules),*] }
        }
    }
}
