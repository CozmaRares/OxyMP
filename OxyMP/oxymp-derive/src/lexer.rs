#![allow(non_snake_case)]

use proc_macro2::{Ident, TokenStream};
use quote::quote;

use crate::{
    data::MacroData,
    idents::{base_ident, tokens::*},
    symbols::{get_def, Symbol},
    tokens::*,
};

pub fn generate_lexer(data: &MacroData) -> TokenStream {
    let tokens = generate_tokens(data);
    let constructor = generate_constructor(data);

    quote! {
        #tokens
        #constructor
    }
}

fn generate_token_idents(token: &TokenInfo) -> Option<(Ident, Ident, Option<Ident>)> {
    match token {
        TokenInfo::Exact(ExactToken { name, .. }) => {
            Some((enum_ident(name), struct_ident(name), None))
        }
        TokenInfo::Regex(RegexToken { name, kind, .. }) => {
            Some((enum_ident(name), struct_ident(name), Some(base_ident(kind))))
        }
        TokenInfo::Ignore(_) => None,
    }
}

fn generate_tokens(data: &MacroData) -> TokenStream {
    let visibility = &data.visibility;
    let token_info = &data.tokens;

    let idents = token_info
        .iter()
        .filter_map(|info| generate_token_idents(info));

    let enum_entries = idents.clone().map(|(enum_entry, struct_ident, _)| {
        quote! {
            #enum_entry(#struct_ident)
        }
    });

    let _Rc = get_def(Symbol::Rc, data.simple_types);
    let _Debug = get_def(Symbol::DeriveDebug, data.simple_types);
    let _Clone = get_def(Symbol::DeriveClone, data.simple_types);

    let structs = idents
        .map(|idents| match idents {
            (_, struct_ident, None) => quote! {
                #visibility struct #struct_ident;
            },
            (_, struct_ident, Some(kind_ident)) => quote! {
                #visibility struct #struct_ident(#visibility #_Rc<#kind_ident>);
            },
        })
        .map(|struct_def| {
            quote! {
                #[derive(#_Debug, #_Clone)]
                #struct_def
            }
        });

    quote! {
        #(#structs)*

        #[derive(#_Debug, #_Clone)]
        #visibility enum Token {
            #(#enum_entries),*
        }
    }
}

fn generate_constructor(data: &MacroData) -> TokenStream {
    let visibility = &data.visibility;
    let token_info = &data.tokens;

    let _Rc = get_def(Symbol::Rc, data.simple_types);
    let _Box = get_def(Symbol::Box, data.simple_types);
    let _vec = get_def(Symbol::VecMacro, data.simple_types);
    let _LexRule = get_def(Symbol::UtilLexRule, data.simple_types);
    let _Lexer = get_def(Symbol::UtilLexer, data.simple_types);
    let _TokenHandler = get_def(Symbol::UtilTokenHandler, data.simple_types);
    let _TokenMatcher = get_def(Symbol::UtilTokenMatcher, data.simple_types);

    let rules = token_info.iter().map(|tok| match tok {
        TokenInfo::Exact(ExactToken { name, pattern }) => {
            let enum_ident = enum_ident(name);
            let struct_ident = struct_ident(name);

            quote! {
                #_LexRule {
                    matcher: #_TokenMatcher::Exact(#pattern.to_string()),
                    handler: #_TokenHandler::Pattern(Token::#enum_ident(#struct_ident))
                }
            }
        }
        TokenInfo::Regex(RegexToken {
            name,
            regex,
            transformer_fn,
            ..
        }) => {
            let enum_ident = enum_ident(name);
            let struct_ident = struct_ident(name);
            let fn_ident = base_ident(transformer_fn);

            quote! {
                #_LexRule {
                    matcher: #_TokenMatcher::regex(#regex),
                    handler: #_TokenHandler::Regex(
                        #_Box::new(
                            |matched| Token::#enum_ident(#struct_ident(#_Rc::new(#fn_ident(matched))))
                        )
                    )
                }
            }
        }
        TokenInfo::Ignore(IgnorePattern { regex }) => quote! {
            #_LexRule {
                matcher: #_TokenMatcher::regex(#regex),
                handler: #_TokenHandler::Ignore
            }
        },
    });

    quote! {
        #visibility fn create_lexer() -> #_Lexer<Token> {
            return #_Lexer { rules: #_vec![#(#rules),*] };
        }
    }
}
