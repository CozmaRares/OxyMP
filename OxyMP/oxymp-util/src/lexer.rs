use std::{collections::HashMap, fmt::Formatter};

use regex::Regex;

#[derive(Debug)]
pub enum TokenMatcher {
    Exact(String),
    Regex(Regex),
}

impl TokenMatcher {
    pub fn regex(re: &str) -> TokenMatcher {
        let re = format!("^{re}");
        let re = Regex::new(&re).unwrap();
        TokenMatcher::Regex(re)
    }
}

pub enum TokenHandler<Token>
where
    Token: std::fmt::Debug,
{
    Pattern(Token),
    Regex(Box<dyn Fn(&str) -> Token>),
    Ignore,
}

impl<Token> std::fmt::Debug for TokenHandler<Token>
where
    Token: std::fmt::Debug,
{
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        match self {
            TokenHandler::Pattern(token) => f.debug_tuple("Pattern").field(token).finish(),
            TokenHandler::Regex(_) => f.debug_tuple("Regex").finish(),
            TokenHandler::Ignore => f.debug_tuple("Ignore").finish(),
        }
    }
}

#[derive(Debug)]
pub struct LexRule<Token>
where
    Token: std::fmt::Debug + Clone,
{
    pub matcher: TokenMatcher,
    pub handler: TokenHandler<Token>,
}

impl<Token> LexRule<Token>
where
    Token: std::fmt::Debug + Clone,
{
    fn matches(&self, input: &str) -> Option<usize> {
        match &self.matcher {
            TokenMatcher::Exact(exact_match) => {
                input.starts_with(exact_match).then_some(exact_match.len())
            }
            TokenMatcher::Regex(re) => re
                .captures(input)
                .and_then(|captures| captures.get(0))
                .map(|matched| matched.end() - matched.start()),
        }
    }

    pub fn consume<'a>(&self, input: &'a str) -> Option<(Option<Token>, &'a str)> {
        self.matches(input).map(|matched_size| {
            let token = match &self.handler {
                TokenHandler::Ignore => None,
                TokenHandler::Pattern(t) => Some(t.clone()),
                TokenHandler::Regex(f) => Some(f(&input[..matched_size])),
            };
            Some((token, &input[matched_size..]))
        })?
    }
}

#[derive(Debug)]
pub struct LexError<'a> {
    pub input: &'a str,
    pub message: String,
}

#[derive(Debug)]
pub struct Lexer<Token>
where
    Token: std::fmt::Debug + Clone,
{
    pub rules: Vec<LexRule<Token>>,
}

impl<Token> Lexer<Token>
where
    Token: std::fmt::Debug + Clone,
{
    pub fn tokenize(self, mut input: &str) -> Result<Vec<Token>, LexError> {
        let mut tokens = Vec::new();
        while !input.is_empty() {
            let mut was_consumed = false;
            for rule in &self.rules {
                if let Some((token, remaining)) = rule.consume(input) {
                    if let Some(token) = token {
                        tokens.push(token);
                    }
                    input = remaining;
                    was_consumed = true;
                    break;
                }
            }
            if !was_consumed {
                return Err(LexError {
                    input,
                    message: "Unknown token".to_string(),
                });
            }
        }
        Ok(tokens)
    }
}

pub struct LexerBuilder<Tier, Token>
where
    Tier: std::hash::Hash + Ord + Default,
    Token: std::fmt::Debug + Clone,
{
    rules: HashMap<Tier, Vec<LexRule<Token>>>,
}

impl<Tier, Token> LexerBuilder<Tier, Token>
where
    Tier: std::hash::Hash + Ord + Default,
    Token: std::fmt::Debug + Clone,
{
    pub fn new() -> Self {
        LexerBuilder {
            rules: HashMap::new(),
        }
    }

    pub fn add_rule(&mut self, rule: LexRule<Token>) -> &Self {
        let tier = Tier::default();
        self.rules.entry(tier).or_default().extend(vec![rule]);
        self
    }

    pub fn add_tiered_rule(&mut self, tier: Tier, rule: LexRule<Token>) -> &Self {
        self.rules.entry(tier).or_default().extend(vec![rule]);
        self
    }

    pub fn build(self) -> Lexer<Token> {
        let mut rules: Vec<_> = self.rules.into_iter().collect();
        rules.sort_by(|(key1, _), (key2, _)| key1.cmp(key2));

        let rules = rules
            .into_iter()
            .map(|(_, rules)| rules)
            .flatten()
            .collect();

        Lexer { rules }
    }
}
