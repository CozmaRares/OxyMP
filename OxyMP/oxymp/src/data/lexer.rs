use quote::ToTokens;

use super::processor::ItemProcessor;

#[derive(Debug)]
pub struct LexerData {
    pub visibility: proc_macro2::TokenStream,
    pub ident: syn::Ident,
    pub skip_patterns: Vec<String>,
}

pub struct LexerProcessor;

impl ItemProcessor<LexerData, syn::ItemStruct> for LexerProcessor {
    fn get_target() -> &'static str {
        "Lexer"
    }

    fn get_expected_variant() -> &'static str {
        "struct"
    }

    fn get_variant(item: &syn::Item) -> Option<syn::ItemStruct> {
        match item {
            syn::Item::Struct(item) => Some(item.clone()),
            _ => None,
        }
    }

    fn extract_data(
        mut item: syn::ItemStruct,
        makrer_attr_idx: usize,
    ) -> syn::Result<(LexerData, syn::Item)> {
        let visibility = item.vis.to_token_stream();
        let ident = item.ident.clone();

        let mut skip_patterns = Vec::new();
        let mut attributes = Vec::new();

        item.attrs.remove(makrer_attr_idx);

        for attr in item.attrs {
            let path = attr.path();

            if path.is_ident("skip") {
                let skip_attr: SkipAttr = attr.parse_args()?;
                skip_patterns.push(skip_attr.0);
            } else {
                attributes.push(attr);
            }
        }

        item.attrs = attributes;

        Ok((
            LexerData {
                ident,
                visibility,
                skip_patterns,
            },
            syn::Item::Struct(item),
        ))
    }
}

struct SkipAttr(String);

impl syn::parse::Parse for SkipAttr {
    fn parse(input: syn::parse::ParseStream) -> syn::Result<SkipAttr> {
        let pattern: syn::LitStr = input.parse()?;

        if !input.is_empty() {
            return Err(syn::Error::new(
                input.span(),
                "Unexpected tokens. Please remove them.",
            ));
        }

        Ok(SkipAttr(pattern.value()))
    }
}
