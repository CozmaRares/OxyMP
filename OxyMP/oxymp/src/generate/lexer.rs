use std::collections::HashMap;

use quote::format_ident;

use crate::{
    automata::{
        dfa::{
            self, StateKind as DFAStateKind, StateTag as DFAStateTag, Transition as DFATransition,
            DFA,
        },
        nfa::{self, combine, NFA},
    },
    data::{
        lexer::LexerData,
        tokens::{TokenPattern, TokensData},
    },
    utils::combine_errors,
};

pub fn generate(
    tokens_data: &TokensData,
    lexers: Vec<(syn::ItemMod, LexerData)>,
) -> syn::Result<Vec<syn::Item>> {
    let mut lexer_cache = LexerCache::new();
    let mut items = Vec::new();
    let mut errors = Vec::new();

    for (item_mod, lexer_data) in lexers {
        let res = generate_one(tokens_data, lexer_data, item_mod, &mut lexer_cache);
        match res {
            Ok(inner_items) => items.extend(inner_items),
            Err(err) => errors.push(err),
        }
    }

    if errors.is_empty() {
        Ok(items)
    } else {
        Err(combine_errors(errors))
    }
}

fn generate_one(
    tokens_data: &TokensData,
    lexer_data: LexerData,
    mut item_mod: syn::ItemMod,
    cache: &mut LexerCache,
) -> syn::Result<Vec<syn::Item>> {
    let mut items = Vec::new();

    let dfa = generate_lexer_dfa(tokens_data, &lexer_data.skip_patterns, cache)?;

    let (error_idents, error_items) = cache.get_error_idents();
    items.extend(error_items);

    let lexer_items = generate_mod(tokens_data, lexer_data, dfa, error_idents)?;
    let (brace, _) = item_mod.content.expect("module should have content");
    item_mod.content = Some((brace, lexer_items));
    items.push(syn::Item::Mod(item_mod));
    Ok(items)
}

struct LexerCache {
    error_idents: Option<ErrorIdents>,
    tokens_nfa: Option<NFA>,
}

impl LexerCache {
    fn new() -> LexerCache {
        LexerCache {
            error_idents: None,
            tokens_nfa: None,
        }
    }

    fn get_error_idents(&mut self) -> (&ErrorIdents, Vec<syn::Item>) {
        let mut items = Vec::new();

        if self.error_idents.is_none() {
            let ErrorItem {
                idents: error_idents,
                items: error_items,
            } = generate_error_ds();
            items.extend(error_items);
            self.error_idents = Some(error_idents);
        }

        (
            self.error_idents
                .as_ref()
                .expect("error idents should be set"),
            items,
        )
    }

    fn get_tokens_nfa(&mut self, tokens_data: &TokensData) -> syn::Result<&NFA> {
        if self.tokens_nfa.is_none() {
            let nfa = generate_tokens_nfa(tokens_data)?;
            self.tokens_nfa = Some(nfa);
        }
        let nfa = self.tokens_nfa.as_ref().expect("tokens nfa should be set");
        Ok(nfa)
    }
}

struct ErrorIdents {
    expected_ident: syn::Ident,
    error_ident: syn::Ident,
}

struct ErrorItem {
    idents: ErrorIdents,
    items: Vec<syn::Item>,
}

#[allow(non_snake_case)]
fn generate_error_ds() -> ErrorItem {
    let LexerExpected = format_ident!("__Lexer_Common_Expected_{}", rand::random::<u64>());
    let LexerError = format_ident!("__Lexer_Common_Error_{}", rand::random::<u64>());

    let expected_enum: syn::ItemEnum = pq! {
        #[derive(Debug)]
        pub enum #LexerExpected {
            Char(char),
            CharRange { start: char, end: char },
        }
    };

    let expected_display: syn::ItemImpl = pq! {
        impl std::fmt::Display for #LexerExpected {
            fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
                let print_char = |f: &mut std::fmt::Formatter<'_>, c: &char| {
                    if c.is_whitespace() {
                        write!(f, "whitespace(code {})", *c as usize)
                    }
                    else {
                        write!(f, "'{}", c)
                    }
                };

                match self {
                    #LexerExpected::Char(c) => print_char(f, c),
                    #LexerExpected::CharRange { start, end } => {
                        print_char(f, start)?;
                        write!(f, "..=")?;
                        print_char(f, end)
                    }
                }
            }
        }
    };

    let error_enum: syn::ItemEnum = pq! {
        // TODO: add location
        // needs refactor of accept fns
        // accept and reject
        #[derive(Debug)]
        pub enum #LexerError {
            Native {
                unexpected_char: char,
                expected: Vec<#LexerExpected>,
            },
            UserTransform(Box<dyn std::error::Error>),
        }
    };

    let error_error: syn::ItemImpl = pq! {
        impl std::error::Error for #LexerError {}
    };

    let error_display: syn::ItemImpl = pq! {
        impl std::fmt::Display for #LexerError {
            fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
                match self {
                    #LexerError::Native {
                        unexpected_char,
                        expected,
                    } => {
                        let expected = expected
                            .iter()
                            .map(|expected| format!("{}", expected))
                            .collect::<Vec<_>>()
                            .join(", ");
                        write!(
                            f,
                            "Unexpected character '{}'\nExpected: {}",
                            unexpected_char, expected
                        )
                    }
                    #LexerError::UserTransform(e) => write!(f, "User transform error: {}", e),
                }
            }
        }
    };

    ErrorItem {
        idents: ErrorIdents {
            expected_ident: LexerExpected,
            error_ident: LexerError,
        },
        items: vec![
            syn::Item::Enum(expected_enum),
            syn::Item::Impl(expected_display),
            syn::Item::Enum(error_enum),
            syn::Item::Impl(error_error),
            syn::Item::Impl(error_display),
        ],
    }
}

fn generate_tokens_nfa(tokens_data: &TokensData) -> syn::Result<NFA> {
    let mut token_nfas = Vec::new();

    for (idx, variant) in tokens_data.variants.iter().enumerate() {
        let pattern = match &variant.pattern {
            TokenPattern::Exact { pattern } => pattern.value(),
            TokenPattern::Regex { pattern, .. } => pattern.value(),
        };

        let nfa = nfa::compile(
            &pattern,
            nfa::StateTag::Token {
                variant: variant.ident.to_string(),
                priority: idx,
            },
        )
        .map_err(|e| {
            syn::Error::new(
                variant.pattern.span(),
                format!(
                    "Error while compiling regex pattern for token variant '{}'\n{}",
                    variant.ident, e
                ),
            )
        })?;

        token_nfas.push(nfa);
    }

    Ok(nfa::combine(token_nfas))
}

fn generate_lexer_dfa(
    tokens_data: &TokensData,
    skip_patterns: &[syn::LitStr],
    cache: &mut LexerCache,
) -> syn::Result<DFA> {
    let token_nfa = cache.get_tokens_nfa(tokens_data)?;

    let mut nfas = Vec::new();

    for pattern_lit in skip_patterns {
        let pattern = pattern_lit.value();
        let nfa = nfa::compile(
            &pattern,
            nfa::StateTag::Skip {
                pattern: pattern.clone(),
            },
        )
        .map_err(|e| {
            syn::Error::new(
                pattern_lit.span(),
                format!(
                    "Error while compiling regex pattern for skip pattern '{}'\n{}",
                    pattern, e
                ),
            )
        })?;

        nfas.push(nfa);
    }

    nfas.push(token_nfa.clone());
    let nfa = nfa::combine(nfas);
    let dfa = dfa::compile(nfa);
    Ok(dfa)
}

fn generate_mod(
    tokens_data: &TokensData,
    lexer_data: LexerData,
    dfa: DFA,
    error_idents: &ErrorIdents,
) -> syn::Result<Vec<syn::Item>> {
    let tokenize_fn = generate_tokenize_fn(lexer_data.visibility);

    let states = generate_states(&dfa);

    let token_transforms = tokens_data
        .variants
        .iter()
        .map(|variant| {
            let variant_ident = variant.ident.clone();
            let ident = variant.ident.to_string();

            let transform = match &variant.pattern {
                TokenPattern::Exact { .. } => q! { Token::#variant_ident },
                TokenPattern::Regex { transform, .. } => q! {
                    super::#transform(inp)
                        .map(Token::#variant_ident)
                        .map_err(|e| Error::UserTransform(Box::new(e)))?
                },
            };

            (ident, transform)
        })
        .collect();

    let methods = generate_state_methods(&dfa, &token_transforms);

    let tokens_ident = &tokens_data.ident;

    let ErrorIdents {
        expected_ident,
        error_ident,
    } = error_idents;

    let items = items! {
        type Token = super::#tokens_ident;
        pub type Error = super::#error_ident;
        pub type LexerExpected = super::#expected_ident;

        #states
        #(#methods)*
        #tokenize_fn
    };
    Ok(items)
}

fn generate_states(dfa: &DFA) -> proc_macro2::TokenStream {
    let idents: Vec<_> = dfa
        .states()
        .iter()
        .map(|(state_id, _)| format_ident!("_{}", state_id))
        .collect();

    let state_structs = idents.iter().map(|ident| q! { struct #ident; });
    let transition_branches = idents
        .iter()
        .map(|ident| q! { State::#ident => #ident::transition(c), });
    let accept_branches = idents
        .iter()
        .map(|ident| q! { State::#ident => #ident::accept(inp, c), });

    q! {
        #[derive(Debug)]
        enum State {
            #(#idents),*
        }

        #(#state_structs)*

        impl State {
            fn transition(&self, c: char) -> Option<State> {
                match self {
                    #(#transition_branches)*
                }
            }

            fn accept(&self, inp: &str, c: char) -> Result<Option<Token>, Error> {
                match self {
                    #(#accept_branches)*
                }
            }
        }
    }
}

fn generate_state_methods<'a>(
    dfa: &'a DFA,
    token_transforms: &'a HashMap<String, proc_macro2::TokenStream>,
) -> impl Iterator<Item = proc_macro2::TokenStream> + 'a {
    dfa.states().iter().map(move |(state_id, state)| {
        let transition_branches = state.transitions.iter().map(|(transition, target)| {
            let target = format_ident!("_{}", target);
            let target = q! { Some(State::#target) };
            match transition {
                DFATransition::Char(c) => q! { #c => #target, },
                DFATransition::Chars { start, end } => q! { #start..=#end => #target, },
            }
        });

        let accept = match &state.kind {
            DFAStateKind::NotAccepting => {
                let expected = state
                    .transitions
                    .iter()
                    .map(|(transition, _)| match transition {
                        DFATransition::Char(c) => q! {
                            LexerExpected::Char(#c)
                        },
                        DFATransition::Chars { start, end } => q! {
                            LexerExpected::CharRange{ start: #start, end: #end }
                        },
                    });

                q! {
                    Err(
                        Error::Native {
                            unexpected_char: current_char,
                            expected: vec![ #(#expected),* ]
                        }
                    )
                }
            }
            DFAStateKind::Accepting(tag) => match tag {
                DFAStateTag::Token { variant, .. } => {
                    let transform = token_transforms
                        .get(variant)
                        .expect("token pattern should always have an transform");
                    q! { Ok(Some(#transform)) }
                }
                DFAStateTag::Skip { .. } => q! { Ok(None) },
            },
        };

        let ident = format_ident!("_{}", state_id);

        q! {
            impl #ident {
                fn transition(c: char) -> Option<State> {
                    match c {
                        #(#transition_branches)*
                        _ => None,
                    }
                }

                fn accept(inp: &str, current_char: char) -> Result<Option<Token>, Error> {
                    #accept
                }
            }
        }
    })
}

fn generate_tokenize_fn(vis: proc_macro2::TokenStream) -> proc_macro2::TokenStream {
    q! {
        #vis fn tokenize(inp: &str) -> Result<Vec<Token>, Error> {
            let mut toks = Vec::new();
            let mut state = State::_1;
            let mut match_start = 0;
            let mut iter = inp.chars().enumerate().peekable();

            while let Some((current_index, c)) = iter.peek() {
                match state.transition(*c) {
                    Some(new_state) => {
                        state = new_state;
                        iter.next();
                    }
                    None => {
                        if let Some(tok) = state.accept(&inp[match_start..*current_index], *c)? {
                            toks.push(tok)
                        }
                        state = State::_1;
                        match_start = *current_index;
                    }
                }
            }

            if inp.len() > 0 {
                if let Some(tok) = state.accept(&inp[match_start..], inp.chars().last().unwrap())? {
                    toks.push(tok)
                }
            }

            Ok(toks)
        }
    }
}
