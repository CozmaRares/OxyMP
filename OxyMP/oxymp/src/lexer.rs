use std::collections::{HashMap, HashSet};

use crate::attribute::{AttributeList, KeyValue};

//#[regex_token(
//    name = "Number",
//    regex = r"-?[0-9]+",
//    transformer_fn = "match_number",
//    kind = "i64"
//)]
//#[ignore_pattern(regex = r"\s+")]

#[derive(Debug)]
pub struct TokenExactMatch {
    name: String,
    pattern: String,
}

#[derive(Debug)]
pub struct TokenRegex {
    name: String,
    regex: String,
    transformer_fn: String,
    kind: String,
}

#[derive(Debug)]
pub struct TokenIgnore {
    regex: String,
}

#[derive(Debug)]
pub enum Token {
    ExactMatch(TokenExactMatch),
    Regex(TokenRegex),
    Ignore(TokenIgnore),
}

impl Token {
    //#[exact_token(name = "Minus", pattern = "-")]
    pub fn exact_token(tokens: proc_macro2::TokenStream) -> syn::Result<Self> {
        let mut expected_properties = HashSet::new();
        expected_properties.insert("name".to_string());
        expected_properties.insert("pattern".to_string());

        let AttributeList { pairs, .. } = AttributeList::prepare_token_info(
            tokens,
            "exact_token".to_string(),
            expected_properties,
        )?;

        let map = pairs
            .iter()
            .fold(HashMap::new(), |mut acc, KeyValue { name, value }| {
                acc.insert(name.content.clone(), value);
                acc
            });

        return Ok(Token::ExactMatch(TokenExactMatch {
            name: map.get("name").unwrap().to_string(),
            pattern: map.get("pattern").unwrap().to_string(),
        }));
    }
}
