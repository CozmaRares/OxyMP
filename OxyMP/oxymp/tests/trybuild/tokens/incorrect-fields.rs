// exact variant's fields must be unit
#[oxymp::oxymp]
mod l1 {
    #[oxymp::Tokens]
    enum Tok {
        #[exact("while")]
        While(i64),
    }
}

// regex variant's fields must not be unit
#[oxymp::oxymp]
mod l2 {
    use oxymp_util::lexer::LexResult;
    fn match_number(_: &str) -> LexResult<Tok> {
        Ok(Tok::Number)
    }

    #[oxymp::Tokens]
    enum Tok {
        #[regex(r"\d+(.\d+)?", match_number)]
        Number,
    }
}

// regex variant with empty tuple
#[oxymp::oxymp]
mod l3 {
    #[oxymp::Tokens]
    enum Tok {
        #[regex(r"\d+(.\d+)?", match_number)]
        Number(),
    }
}

// regex variant with empty struct
#[oxymp::oxymp]
mod l4 {
    #[oxymp::Tokens]
    enum Tok {
        #[regex(r"\d+(.\d+)?", match_number)]
        Number {},
    }
}

fn main() {}
