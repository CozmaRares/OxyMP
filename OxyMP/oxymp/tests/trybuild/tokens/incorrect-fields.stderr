error: Exact tokens can't contain any data. Consider removing any associated data for this token variant.
 --> tests/trybuild/tokens/incorrect-fields.rs:7:14
  |
7 |         While(i64),
  |              ^^^^^

error: Regex tokens must contain some data. Make sure that the variant definition includes encapsulated data.
  --> tests/trybuild/tokens/incorrect-fields.rs:22:9
   |
22 |         Number,
   |         ^^^^^^

error: Regex tokens must contain some data. Make sure that the variant definition includes encapsulated data.
  --> tests/trybuild/tokens/incorrect-fields.rs:32:9
   |
32 |         Number(),
   |         ^^^^^^

error: Regex tokens must contain some data. Make sure that the variant definition includes encapsulated data.
  --> tests/trybuild/tokens/incorrect-fields.rs:42:9
   |
42 |         Number {},
   |         ^^^^^^
